src/lib/RoutingManager.js

```diff
@@ -0,0 +1,195 @@
+    // be not present if it was not going trough the SearchParameters constructor
+    this.originalConfig = algoliasearchHelper(
+      { addAlgoliaAgent() {} },
+      currentConfiguration.index,
+      currentConfiguration
+    ).state;
+    return Object.assign(
+      {},
+      this.getAllSearchParameters({
+        state: this.originalConfig,
+        uiState: this.originalUIState,
+      })
+    );
+  }
+
+  render({ state }) {
+    if (this.firstRender) {
+      this.firstRender = false;
+      const { helper } = this.instantSearchInstance;
+      this.router.onUpdate(route => {
+        const uiState = this.stateMapping.routeToState(route);
+        const currentUIState = this.getAllUIStates({
+          helper,
+          state: helper.state,
+          instantSearchInstance: this.instantSearchInstance,
+        });
+
+        if (isEqual(uiState, currentUIState)) return;
+
+        // retrieve a search state from the widgets and the UI state
+        const searchParameters = this.getAllSearchParameters({
+          state,
+          instantSearchInstance: this.instantSearchInstance,
+          uiState,
+        });
+
+        const fullHelperState = Object.assign(
+          {},
+          this.originalConfig,
+          searchParameters
+        );
+
+        if (isEqual(fullHelperState, searchParameters)) return;
+
+        helper
+          .overrideStateWithoutTriggeringChangeEvent(searchParameters)
+          .search();
+      });
+
+      helper.on('change', searchState => {
+        // this.renderURLFromState(s)
+        const uiState = this.getAllUIStates({
+          helper,
+          state: searchState,
+          instantSearchInstance: this.instantSearchInstance,
+        });
+        const route = this.stateMapping.stateToRoute(uiState);
+        this.router.write(route);
+      });
+
+      // Compare initial state and post first render state, in order
+      // to see if the query has been changed by a searchFunction
+
+      const firstRenderState = this.getAllUIStates({
+        helper,
+        state,
+        instantSearchInstance: this.instantSearchInstance,
+      });
+
+      if (!isEqual(this.initState, firstRenderState)) {
+        // force update the URL, if the state has changed since the initial URL read
+        // We do this in order to make a URL update when there is search function
+        // that prevent the search of the initial rendering
+        // See: https://github.com/algolia/instantsearch.js/issues/2523#issuecomment-339356157
+        this.router.write(firstRenderState);
+      }
+    }
+  }
+
+  dispose() {
+    this.instantSearchInstance.helper.removeListener(
+      'change',
+      this.renderURLFromState
+    );
+    this.router.dispose();
+  }
+
+  getAllSearchParameters({ state, uiState }) {
+    const { widgets } = this.instantSearchInstance;
+    const searchParameters = widgets.reduce((u, w) => {
+      if (!w.getWidgetSearchParameters) return u;
```

Comment:
We can get rid of the condition by doing a filter before the reduce.

Reply:
We're gonna do twice the interations... But it's worth considering.